### PR TITLE
[dnm] kvserver: extra logging for restore/tpce/8TB/aws/nodes=10/cpus=8

### PR DESCRIPTION
**Pull request by tbg**
Touches https://github.com/cockroachdb/cockroach/issues/97019.

### CODE REVIEW

**Comment by tbg** [comment]
TL;DR: I think we're seeing n9 suddenly receive more MsgApp. It gets overwhelmed, the receive queues start filling up (consuming some memory) and there is possibly significant build-up of memory in the RawNodes' unstable raft logs. Additionally, the raft entry cache likely becomes ineffective for log application, meaning we have to start reading from disk at 100mb/s+, which completely starves out the already maxed-out write bandwidth, leading to a quick death spiral.

Some of this is conjecture that will be validated with a next round of experiments that have better metrics added.

----

Notes on investigation. Got a crash, n9 died with an OOM. It had fairly elevated memory usage for many minutes leading up to the crash:

image

The above is RSS, but for our purposes it's more useful to look at the Go heap size:

image

**So we note that n9 suddenly went from a cluster-wide baseline of ~1GB to ~7GB**!

There's a profile from the plateau. It reflects "nodes accounting for 7258.09MB, 100% of 7258.09MB total", i.e. matching very well the observed heap size in the graph. In other words, this is a good profile to look at. Nothing surprising in there, we see what we have seen in other OOMs of the same kind, that entries allocated from incoming MsgApp streams are to blame for the bulk of the consumption (i.e. this is a kind of "follower overload"):

image

So far so good - we want to know who is referencing the memory (i.e. keeping it alive, vs having been the one allocating it in the first place- which is what the heap profile tells us but which isn't relevant here). Hopefully my extra logging can shed some light on this.

The raft receive queue no doubt does fill up, let's call it peak usage of ~1GB.

image

My logging also gives a breakdown on roughly how many replicas contribute to this:

> tobias-1678829488-01-n10cpu8-0009> I230315 01:58:07.429105 325 kv/kvserver/store.go:2884 ⋮ [T1,n9,s9] 8736 recv-queues: ‹r3286: 160 MiB/#56, r3284: 123 MiB/#60, r4783: 120 MiB/#90, r4785: 90 MiB/#44, r2074: 86 MiB/
#36, r4582: 72 MiB/#28, r4784: 50 MiB/#38, r1977: 34 MiB/#12, r2073: 28 MiB/#23, r4093: 20 MiB/#10, r4779: 20 MiB/#82, r4090: 20 MiB/#34, r4091: 20 MiB/#14, r1974: 17 MiB/#63, r4581: 12 MiB/#49, r2660: 12 MiB/#21,
r4092: 10 MiB/#13, r4786: 10 MiB/#17, r4: 487 KiB/#10, [... rest omitted because it gets small...]

which is roughly this distribution below, i.e. we have ~20 receive queues with meaningful data in them.

image

This explains ~1GB of the usage but not the additional 5GB of memory. The raft entry cache, at least according to its metrics, is tiny, too:

image

I suspect that if we looked at the raft entry cache sizing, we'd make it larger. Eyeballing the rates of the `raft.entrycache.{accesses,hits}` against each other, I'd say we have like a >=50% hit rate on this workload. But either way 8mb seems tiny. Besides - it's configured for 16mb[^1] so it's apparently unable to use that, possibly because the entries here are large-ish SSTs. I filed https://github.com/cockroachdb/cockroach/issues/98666 to investigate further.

Where else could the entries be? They could be "inside the RawNode" (in raft's unstable log) waiting to be moved to disk - this is where they go when being pulled out of the receive queue. It's a bit hard to piece it together because the logging I added wasn't optimal. What I have looks like this, on a per-range basis:

```
[...] r2851: unstable log #9/60 MiB
```

Since this is in a hot path (handle raft ready loop), I only log when the size is >=50mb, which is an important limitation. It's hard to get a sense how much unstable log is present in aggregate across all `RawNode`s. But one interesting clue is that the first time I see this message fire is at `01:50:18` which is right around the time the memory usage first elevates on n9, and from that point on it fires at least every couple of seconds.

I am seeing it fire on 76 distinct rangeIDs, typically with unstable logs in the 60-300mb range. So as an overly pessimistic worst-case estimate, let's say all 76 ranges had built up 100mb: that would easily explain 7+GB of memory usage. The reality may be different, but there seems to be promise in this explanation, especially due to the 50mb cutoff.

I'm going to improve the tracking - probably adding an `aggmetric` Gauge so that we can track a running total of the combined unstable logs which we'll hopefully see spike to 4+GB on the next repro.

What this won't explain though is the regime change that seems to happen here. What exactly happens at `1:50` to have n9 leap off the cliff?

Looking at the `cr.store.raft.rcvd.app, cr.store.raft.rcvd.appresp` metrics, we're mostly receiving MsgApp, i.e. we're a follower mostly. Interestingly, the rate seems to elevate right as the memory usage does too:

image

so perhaps there was some range movement, or other event that now caused significantly more traffic to flow to n9, overwhelming it. (This does not explain the reads that then start and probably made it a lot worse, though).

We see disk _reads_ spike a little after the first elevation of memory on n9:

image

which would certainly explain something, since these machines are under I/O throttling and read bandwidth is shared with write bandwidth (note the ceiling at ~240mb/s in the graph below, which n9 (brown) is constantly pushing against):

image

I checked compaction read bytes and outgoing snapshots; they're not it. It's unlikely to be consistency checks either since those are rate limited and I don't see a pattern change in their metrics.

The best conjecture I have is that by being overloaded, the node eventually falls behind on command application because we allow at most 64MiB to be applied in a single cycle. As the spread between last index and appliable index goes up, the efficiency of the raft entry cache goes down - we end up having to pull the entries we are applying from disk nearly 100% of the time. Note how in the graph below, the hit rate for the cache drops at around `1:55`, which is when the disk reads increase.

image

The bytes pulled from `raft.Storage` are something we should have a metric for too, which I will add for the next round of experiments.

[^1]: https://github.com/cockroachdb/cockroach/blob/e005c0f90ccc3b5462b9ff132020f1257d7ea53d/pkg/kv/kvserver/store.go#L103-L105

**Comment by tbg** [comment]
New round. Not a failure, but a pronounced spike. https://pprof.me/3774cb2/

Showing nodes accounting for 3476.78MB, 94.81% of 3667.08MB total

Incoming MsgApp:

image

Loading entries for application:

image

from AddSST application (there's an extra level of unmarshalling) and StoreEntries (via MaybeSideloadEntries)

image

I'm focusing on the incoming MsgApp right now, so we're looking to find out where 2GB of entries are being held.


<details>
<summary>~480mb in the raft receive queue</summary>

image


</details>

<details>
<summary>~1.68gb in the unstable logs</summary>

image

</details>

It almost feels too good, those sum up pretty much exactly to what's represented in the heap profile. So we seem to understand very well that there can be significant amount of build-up both in the receive queues but in particular in the unstable log when a follower gets overloaded and its raft handling slows down (in this case due to slow disk).

I don't see any of my per-replica "unstable log" messages with a size of >>100mb (though I've seen ~300mb in a previous run, so it does happen when things get worse) but see many different replicas hit the >50mb threshold (which is required for being logged), so it's more of a steady build-up across probably a dozen or two replicas.

For the raft queues, like in the previous investigation we see like 10-20 replicas with substantial build-up.


```
[T1,n10,s10] 7263  recv-queues: total 552 MiB, details: ‹r2448: 104 MiB/#52, r5439: 103 MiB/#56, r2446: 98 MiB/#53, r2461: 60 MiB/#30, r2462: 52 MiB/#14, r5358: 48 MiB/#33, r5357: 40 MiB/#47, r5441: 16 MiB/#13, r5355: 15 MiB/#40, r2165: 8.7 MiB/#4, r2163: 8.6 MiB/#5, [... <10kb ...]
```

----

As for the other build-up, inlining SSTs for log application only happens when we need to pull entries from disk, so better sizing of the raft entry cache can help - or a way to avoid ever pulling the payloads into memory in the first place - we never care about the SST in memory, as we link it into the LSM directly from the file system. There is potential for eliding much of this memory use in the first place - this would elide both other sources of inuse_mem entirely, i.e. remove both of these:

image

**Comment by tbg** [comment]
Writing this for the third time - `pprof.me` reliably crashes my browser when uploading the heap profile.

I ran the experiment again, inserting disk slowness artificially[^1] (got impatient) on n5. It worked well: RSS spiked to 12GiB, unstable log peaked at 3.44GiB, receive queue at 1.22GiB.

[^1]: https://github.com/cockroachdb/cockroach/pull/81516/files
